### PR TITLE
feat: add built-in types

### PR DESCRIPTION
**Pull request by leanmendoza**
# Whats?
Add Enum<T>, Boolean and Optional types.

### CODE REVIEW

File: src/built-in-types/MapType.ts [hunk 1]
```diff
@@ -5,27 +5,42 @@ export interface Spec {
...
+
+type OnlyOptionalUndefinedTypes<T> = { [K in IncludeUndefined<T>]?: T[K] }
+type OnlyNonUndefinedTypes<T> = { [K in ExcludeUndefined<T>]: T[K] }
+
+type ToOptional<T> = OnlyOptionalUndefinedTypes<T> & OnlyNonUndefinedTypes<T>
+
+export type Result<T extends Spec> = ToOptional<{
   [K in keyof T]: T[K] extends EcsType
-    ? ReturnType<T[K]['deserialize']>
+    ? ReturnType<T[K]['deserialize']> extends undefined
```
Comment (gonpombo8) [inline src/built-in-types/MapType.ts:25]: i think there is no need for this change. 🤔

File: src/built-in-types/MapType.ts [hunk 2]
```diff
@@ -5,27 +5,42 @@ export interface Spec {
   [key: string]: EcsType
 }
 
-export type Result<T extends Spec> = {
+/** Include property keys from T where the property is assignable to U */
+type IncludeUndefined<T> = {
```
Comment (gonpombo8) [inline src/built-in-types/MapType.ts:9]: can we move all this types to a types.ts file ?

File: src/built-in-types/MapType.ts [hunk 3]
```diff
@@ -5,27 +5,42 @@ export interface Spec {
...
-}
+}>
 
 export function MapType<T extends Spec>(spec: T): EcsType<Result<T>> {
   return {
     serialize(value: Result<T>, builder: ByteBuffer): void {
       for (const key in spec) {
-        const type = spec[key]
-        type.serialize(value[key], builder)
+        spec[key].serialize((value as any)[key], builder)
```
Comment (gonpombo8) [inline src/built-in-types/MapType.ts:37]: can we create a ticket for this `as any` ?

File: src/built-in-types/MapType.ts (hunk 2 above)
Comment (leanmendoza) [reply src/built-in-types/MapType.ts]: done

File: src/built-in-types/MapType.ts (hunk 1 above)
Comment (leanmendoza) [reply src/built-in-types/MapType.ts]: thx

File: src/built-in-types/MapType.ts (hunk 3 above)
Comment (leanmendoza) [reply src/built-in-types/MapType.ts:22]: I added a TODO comment a I'm creating a ticket

File: src/built-in-types/typing.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+/** Excludes property keys from T where the property is assignable to U */
+type ExcludeUndefined<T> = {
+  [P in keyof T]: undefined extends T[P] ? never : P
+}[keyof T]
+
+type OnlyOptionalUndefinedTypes<T> = { [K in IncludeUndefined<T>]?: T[K] }
+type OnlyNonUndefinedTypes<T> = { [K in ExcludeUndefined<T>]: T[K] }
+
+export type ToOptional<T> = OnlyOptionalUndefinedTypes<T> &
+  OnlyNonUndefinedTypes<T>
```
Comment (menduz) [inline src/built-in-types/typing.ts:15]: 🧙

**Pull request merged**